File: 03.function_annotation/bin/uniprot/cog.data.R
args=commandArgs(TRUE)
if (length(args) != 4) {
  print ("usage:Rscript cog.data.R <eggnog.txt> <cog.txt> <CogClass.txt> <Outfile>")
  q()
}
inputFile <- args[1]
inref1 <- args[2]   ##"/calculate/database/uniprot/2022-03-09/cog.txt"
inref2 <- args[3]   ##"/calculate/database/uniprot/2022-03-09/CogClass.txt"
outfile <- args[4]
a <- as.data.frame(read.delim(file=inputFile, header = F,check.names = F,encoding = "UTF-8",comment.char = '#'))[,c(1,5)]
ref1 <- as.data.frame(read.delim(file = inref1, header = F, row.names = 1,check.names = F,encoding = "UTF-8"))
ref2 <- as.data.frame(read.delim(file = inref2, header = F, row.names = 1,check.names = F,encoding = "UTF-8"))
# ref1$CogClassName <- ref1$CogName
# ref1$CogClassName[ref1$CogClassCode==rownames(ref2)] <- ref2[rownames(ref2),]

colnames(a) <- c("query","eggNOG_OGs")
colnames(ref1) <- c("CogClassCode","CogName")
colnames(ref2) <- c("CogClassName")
for (i in 1:nrow(a)) {
  a$eggNOG_OGs[i] <- unlist(strsplit(a$eggNOG_OGs[i],"@"))[1]
}
# a$eggNOG_OGs <- unlist(strsplit(a$eggNOG_OGs,"@"))

a <- a[a$eggNOG_OGs %in% rownames(ref1),]


a$CogName<-a$eggNOG_OGs
a$CogClassName<-a$eggNOG_OGs
a$CogClassCode<-a$eggNOG_OGs
colnames(a) <- c("SeqID","CogID","CogName","CogClassName","CogClassCode")


# b<-data.frame(ncol=c("SeqID","CogID","CogName","CogClassName","CogClassCode"))
b<-a

for (i in 1:nrow(b)) {
  a$CogName[i] <-  ref1$CogName[rownames(ref1)==a$CogID[i]]
  a$CogClassCode[i] <-  ref1$CogClassCode[rownames(ref1)==a$CogID[i]]
  type <- strsplit(a$CogClassCode[i],"")[[1]]  
  if(length(type)>1){
    a$CogClassCode[i] <- type[1]
    a$CogClassName[i] <- ref2$CogClassName[rownames(ref2)==type[1]]
    for (y in 2:length(type)) {
      x<- unlist(c(a[i,1:3],ref2$CogClassName[rownames(ref2)==type[y]],type[y]))
      a <- rbind(a,x)
    }
  }else{
    a$CogClassName[i] <- ref2$CogClassName[rownames(ref2)==type[1]]
    
  }
  
}

write.table(a,file = outfile,sep = "\t",quote = F,fileEncoding = "utf-8",row.names = F)

File: 03.function_annotation/bin/uniprot/cogData.py
import sys,os

if len(sys.argv) != 5:
	print("python %s tmp.emapper.annotations /annotation/uniprot/cog.txt /annotation/uniprot/CogClass.txt cog.xls" %sys.argv[0])
	exit(1)

with open(sys.argv[3], 'r') as f : lines = [i.strip().split('\t') for i in f.readlines()]
cogClass = {CogClassCode:CogClassName for CogClassCode,CogClassName in lines}

with open(sys.argv[2], 'r', encoding='utf-8') as f : lines = [i.strip().split('\t') for i in f.readlines()]
cog = {}
for line in lines:
	cogId,CogClassCode,CogName = line
	cog[cogId] = (CogClassCode,CogName)
if os.path.basename(sys.argv[2]).startswith('cog'):
	out = 'SeqID\tCogID\tCogName\tCogClassName\tCogClassCode\n'
else:
	out = 'SeqID\tKogID\tKogName\tKogClassName\tKogClassCode\n'
with open(sys.argv[1], 'r') as f :
	for line in f:
		if "#" in line : continue
		lst = line.split('\t')
		seqid = lst[0]
		cogId = lst[4].split('@')[0]
		if not cogId or not cogId in cog : continue
		for code in cog[cogId][0]:
			out += '%s\t%s\t%s\t%s\t%s\n' %(seqid,cogId,cog[cogId][1],cogClass[code],code)

with open(sys.argv[4], 'w', encoding='utf-8') as w : w.write(out)
